### DISCUSSION TITLE
Does relay client transport use QUIC?

### DISCUSSION THREAD

**Discussion by mycognosist** [libp2p/rust-libp2p#3766] [Q&A] [answered]
In a previous discussion question, @thomaseizinger helped me to understand [How to combine QUIC and relay client transports? #3653](https://github.com/libp2p/rust-libp2p/discussions/3653).

I have a follow-up question: does the relay transport in the example use QUIC under the hood? Or is it using some other connection protocol?

```rust
let quic_config = quic::Config::new(key_pair);
let quic_transport = quic::tokio::Transport::new(quic_config);

let (relay_transport, relay_client) = relay::client::new(key_pair.public().to_peer_id());
let relay_client = Some(relay_client);

let relay_transport = relay_transport
    .upgrade(Version::V1)
    .authenticate(NoiseAuthenticated::xx(key_pair).unwrap())
    .multiplex(YamuxConfig::default());

let transport = OrTransport::new(quic_transport, relay_transport)
    .map(|either_output, _| match either_output {
        Either::Left((peer_id, muxer)) => (peer_id, StreamMuxerBox::new(muxer)),
        Either::Right((peer_id, muxer)) => (peer_id, StreamMuxerBox::new(muxer)),
    })
    .boxed();
```

**Comment by thomaseizinger** [comment, marked as answer]
Short answer: Yes.

In that code snippet, you only configured relay and QUIC.

This means you can only connect to (and listen on) relays that are reachable via QUIC.

Relay itself isn't a transport. A relayed connection will always run on the connection that you establish to the relay. If the relay only offers TCP+TLS+yamux and your app can only do QUIC, you won't be able to connect to the relay in the first place.

Does that make sense?

**Reply by mycognosist** [reply]
That makes perfect sense. Thanks once again, @thomaseizinger !

I see there's a note about this in the relay transport docs (which I missed previously): https://docs.rs/libp2p/latest/libp2p/relay/client/struct.Transport.html

**Reply by thomaseizinger** [reply]
We have an open issue to add an example for this: https://github.com/libp2p/rust-libp2p/issues/3143

Contributions welcome :)

**Comment by sandreae** [comment]
Hi @mycognosist and @thomaseizinger, thanks for raising and answering this question. I'm also trying to understand this particular case and must admit I'm still a little confused. @thomaseizinger when you say the given example will configure a relay which is only reachable via QUIC transport, I can't find what the passed transport encryption (noise) and multiplexing (yamux) used for... Are they required for cases where we _want_ to only offer relaying over QUIC transport? Or can they be omitted (i understand from this comment [this](https://github.com/libp2p/rust-libp2p/discussions/3653#discussioncomment-5392023) isn't possible though)?

**Reply by mxinden** [reply]
> I can't find what the passed transport encryption (noise) and multiplexing (yamux) used for... 

QUIC itself brings its own encryption and multiplexing. Though a relayed connection (potentially on top of a QUIC connection) needs to provide its own encryption and multiplexing. Thats why we need Noise and Yamux.

**Reply by sandreae** [reply]
Thanks for replying. So although QUIC has its own multiplexing and encryption, we still need Noise and Yamux for configuring the relay (I assume only in the case of other transports being used). This is counterintuitive in our case as we only intend on supporting QUIC across our libp2p stack. Is there anyway to get around needing these dependencies for setting up the relay?

**Reply by sandreae** [reply]
Phrased another way, what are Noise and Yamux used for when the underlying transport is QUIC?

**Reply by thomaseizinger** [reply]
> Thanks for replying. So although QUIC has its own multiplexing and encryption, we still need Noise and Yamux for configuring the relay (I assume only in the case of other transports being used). This is counterintuitive in our case as we only intend on supporting QUIC across our libp2p stack. Is there anyway to get around needing these dependencies for setting up the relay?

You don't need noise and yamux necessarily. It is perfectly fine to deploy a relay with only the QUIC transport. You can additionally also enable a TCP+noise+yamux stack which makes the relay available on both, QUIC and TCP connections.

**Reply by thomaseizinger** [reply]
You could even add in TLS as an alternative to noise if you wanted to.

**Reply by mxinden** [reply]
> > Thanks for replying. So although QUIC has its own multiplexing and encryption, we still need Noise and Yamux for configuring the relay (I assume only in the case of other transports being used). This is counterintuitive in our case as we only intend on supporting QUIC across our libp2p stack. Is there anyway to get around needing these dependencies for setting up the relay?
> 
> You don't need noise and yamux necessarily. It is perfectly fine to deploy a relay with only the QUIC transport. You can additionally also enable a TCP+noise+yamux stack which makes the relay available on both, QUIC and TCP connections.

For the sake of completeness, you need *some* independent security protocol and multiplexer to secure and multiplex the *relayed* connection. E.g. you could establish a relayed connection over two QUIC connections and then upgrade the relayed connection with Noise and Yamux.
